# A191 Stock Factors 141-160

In [3]:
import dai
import pandas as pd
from a191_sql_base import get_a191_sql_base

In [4]:
sd = '2023-01-01'
ed = '2026-01-01'

## A191 Factor 141

Formula:

`(RANK(CORR(RANK(HIGH), RANK(MEAN(VOLUME,15)), 9))* -1)`


In [5]:
sql_alpha_141 = f"""--sql
SELECT
    date,
    instrument,
    (c_pct_rank(m_corr(c_pct_rank(high), c_pct_rank(m_avg(volume, 15)), 9)) * -1)
    AS alpha_a191_f0141
FROM data_base
"""
dai.query(get_a191_sql_base(sql_alpha_141), filters={'date': [sd, ed]}).df().dropna()

date instrument  alpha_a191_f0141
116685  2023-02-10  000001.SZ         -0.462644
116686  2023-02-10  000002.SZ         -0.214286
116687  2023-02-10  000004.SZ         -0.018268
116688  2023-02-10  000005.SZ         -0.433498
116689  2023-02-10  000006.SZ         -0.293719
...            ...        ...               ...
3858340 2025-12-26  688799.SH         -0.616340
3858341 2025-12-26  688800.SH         -0.782457
3858345 2025-12-26  688819.SH         -0.751795
3858346 2025-12-26  688981.SH         -0.848438
3858347 2025-12-26  689009.SH         -0.347953

[3552296 rows x 3 columns]

## A191 Factor 142

Formula:

`(((-1 * RANK(TSRANK(CLOSE, 10))) * RANK(DELTA(DELTA(CLOSE, 1), 1))) *RANK(TSRANK((VOLUME/MEAN(VOLUME,20)), 5)))`


In [6]:
sql_alpha_142 = f"""--sql
SELECT
    date,
    instrument,
    (((-1 * c_pct_rank(m_rank(close, 10))) *
      c_pct_rank(m_delta(m_delta(close, 1), 1))) *
      c_pct_rank(m_rank((volume / m_avg(volume, 20)), 5)))
    AS alpha_a191_f0142
FROM data_base
"""
dai.query(get_a191_sql_base(sql_alpha_142), filters={'date': [sd, ed]}).df().dropna()

date instrument  alpha_a191_f0142
54118   2023-01-17  600866.SH         -0.006408
54120   2023-01-17  600868.SH         -0.056491
54122   2023-01-17  600871.SH         -0.129709
59187   2023-01-18  600866.SH         -0.032868
59189   2023-01-18  600868.SH         -0.344705
...            ...        ...               ...
3858340 2025-12-26  688799.SH         -0.018887
3858341 2025-12-26  688800.SH         -0.005148
3858345 2025-12-26  688819.SH         -0.445836
3858346 2025-12-26  688981.SH         -0.011880
3858347 2025-12-26  689009.SH         -0.008052

[3547049 rows x 3 columns]

## A191 Factor 143

Formula:

`CLOSE>DELAY(CLOSE,1)?(CLOSE-DELAY(CLOSE,1))/DELAY(CLOSE,1)*SELF:SELF`


In [7]:
sql_alpha_143 = f"""--sql
SELECT
    date,
    instrument,
    m_product(IF(close / m_lag(close, 1) > 1, close / m_lag(close, 1), 1), 252)
    AS alpha_a191_f0143
FROM data_base
"""
dai.query(get_a191_sql_base(sql_alpha_143), filters={'date': [sd, ed]}).df().dropna()

date instrument  alpha_a191_f0143
1314363 2024-01-16  000001.SZ          2.860425
1314364 2024-01-16  000002.SZ          3.224274
1314366 2024-01-16  000005.SZ          6.818832
1314367 2024-01-16  000006.SZ          7.217237
1314369 2024-01-16  000008.SZ          3.157426
...            ...        ...               ...
3858340 2025-12-26  688799.SH         25.974338
3858341 2025-12-26  688800.SH        152.103622
3858345 2025-12-26  688819.SH          6.121863
3858346 2025-12-26  688981.SH         13.623877
3858347 2025-12-26  689009.SH         10.881937

[2356340 rows x 3 columns]

## A191 Factor 144

Formula:

`SUMIF(ABS(CLOSE/DELAY(CLOSE,1)-1)/AMOUNT,20,CLOSE<DELAY(CLOSE,1))/COUNT(CLOSE<DELAY(CLOSE,1),20)`


In [8]:
sql_alpha_144 = f"""--sql
SELECT
    date,
    instrument,
    m_sum(
        IF(close < m_lag(close, 1),
           abs(close / m_lag(close, 1) - 1) / amount,
           0),
        20
    ) /
    m_sum(
        IF(close < m_lag(close, 1), 1, 0),
        20
    )
    AS alpha_a191_f0144
FROM data_base
"""
dai.query(get_a191_sql_base(sql_alpha_144), filters={'date': [sd, ed]}).df().dropna()

date instrument  alpha_a191_f0144
101438  2023-02-07  000001.SZ      7.542938e-12
101439  2023-02-07  000002.SZ      1.025348e-11
101440  2023-02-07  000004.SZ      1.181459e-09
101441  2023-02-07  000005.SZ      9.227344e-10
101442  2023-02-07  000006.SZ      4.882287e-11
...            ...        ...               ...
3858340 2025-12-26  688799.SH      2.052943e-10
3858341 2025-12-26  688800.SH      4.058024e-11
3858345 2025-12-26  688819.SH      9.407651e-11
3858346 2025-12-26  688981.SH      3.046133e-12
3858347 2025-12-26  689009.SH      3.946675e-11

[3568142 rows x 3 columns]

## A191 Factor 145

Formula:

`(MEAN(VOLUME,9)-MEAN(VOLUME,26))/MEAN(VOLUME,12)*100`


In [9]:
sql_alpha_145 = f"""--sql
SELECT
    date,
    instrument,
    (m_avg(volume, 9) - m_avg(volume, 26)) / m_avg(volume, 12) * 100
    AS alpha_a191_f0145
FROM data_base
"""
dai.query(get_a191_sql_base(sql_alpha_145), filters={'date': [sd, ed]}).df().dropna()

date instrument  alpha_a191_f0145
131944  2023-02-15  000001.SZ         -4.970611
131945  2023-02-15  000002.SZ         -0.850897
131946  2023-02-15  000004.SZ         -3.079252
131947  2023-02-15  000005.SZ        -11.513002
131948  2023-02-15  000006.SZ        -11.551348
...            ...        ...               ...
3858340 2025-12-26  688799.SH        -16.533971
3858341 2025-12-26  688800.SH          3.107132
3858345 2025-12-26  688819.SH        -14.617498
3858346 2025-12-26  688981.SH          8.726077
3858347 2025-12-26  689009.SH        -19.554688

[3536431 rows x 3 columns]

## A191 Factor 146

Formula:

`MEAN((CLOSE-DELAY(CLOSE,1))/DELAY(CLOSE,1)-SMA((CLOSE-DELAY(CLOSE,1))/DELAY(CLOSE,1),61,2),20)*((CLOSE-DELAY(CLOSE,1))/DELAY(CLOSE,1)-SMA((CLOSE-DELAY(CLOSE,1))/DELAY(CLOSE,1),61,2))/SMA(((CLOSE-DELAY(CLOSE,1))/DELAY(CLOSE,1)-((CLOSE-DELAY(CLOSE,1))/DELAY(CLOSE,1)-SMA((CLOSE-DELAY(CLOSE,1))/DELAY(CLOSE,1),61,2)))^2,60)`


In [10]:
sql_alpha_146 = f"""--sql
SELECT
    date,
    instrument,
    m_avg(
        ((close - m_lag(close, 1)) / m_lag(close, 1) -
         m_ta_ewm((close - m_lag(close, 1)) / m_lag(close, 1), 61, 2)),
        20
    ) *
    ((close - m_lag(close, 1)) / m_lag(close, 1) -
     m_ta_ewm((close - m_lag(close, 1)) / m_lag(close, 1), 61, 2)) /
    m_ta_ewm(
        ((close - m_lag(close, 1)) / m_lag(close, 1) -
         ((close - m_lag(close, 1)) / m_lag(close, 1) -
          m_ta_ewm((close - m_lag(close, 1)) / m_lag(close, 1), 61, 2))) ^ 2,
        60,
        2
    )
    AS alpha_a191_f0146
FROM data_base
"""
dai.query(get_a191_sql_base(sql_alpha_146), filters={'date': [sd, ed]}).df().dropna()

date instrument  alpha_a191_f0146
106518  2023-02-08  000001.SZ          1.604657
106519  2023-02-08  000002.SZ          0.811921
106520  2023-02-08  000004.SZ         -0.348293
106521  2023-02-08  000005.SZ          0.753606
106522  2023-02-08  000006.SZ          0.640750
...            ...        ...               ...
3858340 2025-12-26  688799.SH         -1.854029
3858341 2025-12-26  688800.SH         -8.002584
3858345 2025-12-26  688819.SH          5.222380
3858346 2025-12-26  688981.SH         -2.604402
3858347 2025-12-26  689009.SH         -9.816798

[3562874 rows x 3 columns]

## A191 Factor 147

Formula:

`REGBETA(MEAN(CLOSE,12),SEQUENCE(12))`


In [11]:
sql_alpha_147 = f"""--sql
SELECT
    date,
    instrument,
    m_regr_slope(m_avg(close, 12), row_number(12), 12)
    AS alpha_a191_f0147
FROM data_base
"""
dai.query(get_a191_sql_base(sql_alpha_147), filters={'date': [sd, ed]}).df().dropna()

date instrument  alpha_a191_f0147
116685  2023-02-10  000001.SZ         -2.733540
116686  2023-02-10  000002.SZ        -14.127558
116687  2023-02-10  000004.SZ          0.233660
116688  2023-02-10  000005.SZ         -0.027949
116689  2023-02-10  000006.SZ         -0.308271
...            ...        ...               ...
3858340 2025-12-26  688799.SH          0.079858
3858341 2025-12-26  688800.SH          0.821393
3858345 2025-12-26  688819.SH          0.101903
3858346 2025-12-26  688981.SH          0.219210
3858347 2025-12-26  689009.SH         -0.050187

[3552296 rows x 3 columns]

## A191 Factor 148

Formula:

`((RANK(CORR((OPEN), SUM(MEAN(VOLUME,60), 9), 6)) < RANK((OPEN -TSMIN(OPEN, 14)))) * -1)`


In [12]:
sql_alpha_148 = f"""--sql
SELECT
    date,
    instrument,
    IF(
        (c_pct_rank(m_corr((open), m_sum(m_avg(volume, 60), 9), 6)) <
         c_pct_rank((open - m_min(open, 14)))),
        1,
        0
    ) * -1
    AS alpha_a191_f0148
FROM data_base
"""
dai.query(get_a191_sql_base(sql_alpha_148), filters={'date': [sd, ed]}).df().dropna()

date instrument  alpha_a191_f0148
5067    2023-01-04  000001.SZ               0.0
5068    2023-01-04  000002.SZ               0.0
5069    2023-01-04  000004.SZ               0.0
5070    2023-01-04  000005.SZ               0.0
5071    2023-01-04  000006.SZ               0.0
...            ...        ...               ...
3858343 2025-12-26  688805.SH               0.0
3858344 2025-12-26  688807.SH               0.0
3858345 2025-12-26  688819.SH              -1.0
3858346 2025-12-26  688981.SH              -1.0
3858347 2025-12-26  689009.SH              -1.0

[3668803 rows x 3 columns]

## A191 Factor 149

Formula:

`REGBETA(FILTER(CLOSE/DELAY(CLOSE,1)-1,BANCHMARKINDEXCLOSE<DELAY(BANCHMARKINDEXCLOSE,1)),FILTER(BANCHMARKINDEXCLOSE/DELAY(BANCHMARKINDEXCLOSE,1)-1,BANCHMARKINDEXCLOSE<DELAY(BANCHMARKINDEXCLOSE,1)),252)`


In [13]:
sql_alpha_149 = f"""--sql
SELECT
    date,
    instrument,
    m_regr_slope(
        (close / m_lag(close, 1) - 1),
        (bm_close / m_lag(bm_close, 1) - 1),
        252
    )
    AS alpha_a191_f0149
FROM data_base
"""
dai.query(get_a191_sql_base(sql_alpha_149), filters={'date': [sd, ed]}).df().dropna()

date instrument  alpha_a191_f0149
1319704 2024-01-17  000001.SZ          1.195419
1319705 2024-01-17  000002.SZ          1.141888
1319707 2024-01-17  000005.SZ          0.485554
1319708 2024-01-17  000006.SZ          0.695232
1319710 2024-01-17  000008.SZ          0.797091
...            ...        ...               ...
3858340 2025-12-26  688799.SH          1.346097
3858341 2025-12-26  688800.SH          2.355948
3858345 2025-12-26  688819.SH          1.451088
3858346 2025-12-26  688981.SH          1.590202
3858347 2025-12-26  689009.SH          1.427555

[2351174 rows x 3 columns]

## A191 Factor 150

Formula:

`(CLOSE+HIGH+LOW)/3*VOLUME`


In [14]:
sql_alpha_150 = f"""--sql
SELECT
    date,
    instrument,
    (close + high + low) / 3 * volume
    AS alpha_a191_f0150
FROM data_base
"""
dai.query(get_a191_sql_base(sql_alpha_150), filters={'date': [sd, ed]}).df().dropna()

date instrument  alpha_a191_f0150
5067    2023-01-04  000001.SZ      3.523548e+11
5068    2023-01-04  000002.SZ      3.533492e+11
5069    2023-01-04  000004.SZ      6.608242e+07
5070    2023-01-04  000005.SZ      7.898286e+07
5071    2023-01-04  000006.SZ      3.321986e+10
...            ...        ...               ...
3858343 2025-12-26  688805.SH      5.322368e+08
3858344 2025-12-26  688807.SH      1.078379e+09
3858345 2025-12-26  688819.SH      1.157668e+08
3858346 2025-12-26  688981.SH      4.936753e+09
3858347 2025-12-26  689009.SH      5.856529e+08

[3668803 rows x 3 columns]

## A191 Factor 151

Formula:

`SMA(CLOSE-DELAY(CLOSE,20),20,1)`


In [15]:
sql_alpha_151 = f"""--sql
SELECT
    date,
    instrument,
    m_ta_ewm(close - m_lag(close, 20), 20, 1)
    AS alpha_a191_f0151
FROM data_base
"""
dai.query(get_a191_sql_base(sql_alpha_151), filters={'date': [sd, ed]}).df().dropna()

date instrument  alpha_a191_f0151
106518  2023-02-08  000001.SZ        -31.902139
106519  2023-02-08  000002.SZ       -271.333712
106520  2023-02-08  000004.SZ          1.381713
106521  2023-02-08  000005.SZ         -0.463380
106522  2023-02-08  000006.SZ        -26.868288
...            ...        ...               ...
3858340 2025-12-26  688799.SH         -2.420436
3858341 2025-12-26  688800.SH         10.585530
3858345 2025-12-26  688819.SH          0.359536
3858346 2025-12-26  688981.SH         -0.977942
3858347 2025-12-26  689009.SH         -0.979284

[3562874 rows x 3 columns]

## A191 Factor 152

Formula:

`SMA(MEAN(DELAY(SMA(DELAY(CLOSE/DELAY(CLOSE,9),1),9,1),1),12)-MEAN(DELAY(SMA(DELAY(CLOSE/DELAY(CLOSE,9),1),9,1),1),26),9,1)`


In [16]:
sql_alpha_152 = f"""--sql
SELECT
    date,
    instrument,
    m_ta_ewm(
        m_avg(
            m_lag(
                m_ta_ewm(m_lag(close / m_lag(close, 9), 1), 9, 1),
                1
            ),
            12
        ) -
        m_avg(
            m_lag(
                m_ta_ewm(m_lag(close / m_lag(close, 9), 1), 9, 1),
                1
            ),
            26
        ),
        9,
        1
    )
    AS alpha_a191_f0152
FROM data_base
"""
dai.query(get_a191_sql_base(sql_alpha_152), filters={'date': [sd, ed]}).df().dropna()

date instrument  alpha_a191_f0152
187995  2023-03-02  000001.SZ         -0.026573
187996  2023-03-02  000002.SZ         -0.001067
187997  2023-03-02  000004.SZ         -0.003052
187998  2023-03-02  000005.SZ         -0.005732
187999  2023-03-02  000006.SZ          0.038640
...            ...        ...               ...
3858340 2025-12-26  688799.SH          0.006173
3858341 2025-12-26  688800.SH          0.028880
3858345 2025-12-26  688819.SH         -0.012233
3858346 2025-12-26  688981.SH          0.002704
3858347 2025-12-26  689009.SH          0.015258

[3478287 rows x 3 columns]

## A191 Factor 153

Formula:

`(MEAN(CLOSE,3)+MEAN(CLOSE,6)+MEAN(CLOSE,12)+MEAN(CLOSE,24))/4`


In [17]:
sql_alpha_153 = f"""--sql
SELECT
    date,
    instrument,
    (m_avg(close, 3) + m_avg(close, 6) + m_avg(close, 12) + m_avg(close, 24)) / 4
    AS alpha_a191_f0153
FROM data_base
"""
dai.query(get_a191_sql_base(sql_alpha_153), filters={'date': [sd, ed]}).df().dropna()

date instrument  alpha_a191_f0153
121771  2023-02-13  000001.SZ       1624.267498
121772  2023-02-13  000002.SZ       3088.005194
121773  2023-02-13  000004.SZ         41.166497
121774  2023-02-13  000005.SZ         17.036943
121775  2023-02-13  000006.SZ        218.844319
...            ...        ...               ...
3858340 2025-12-26  688799.SH         73.420755
3858341 2025-12-26  688800.SH        150.819154
3858345 2025-12-26  688819.SH         36.276818
3858346 2025-12-26  688981.SH        118.914583
3858347 2025-12-26  689009.SH         60.031919

[3547007 rows x 3 columns]

## A191 Factor 154

Formula:

`(((VWAP -MIN(VWAP, 16))) < (CORR(VWAP, MEAN(VOLUME,180), 18)))`


In [18]:
sql_alpha_154 = f"""--sql
SELECT
    date,
    instrument,
    IF((vwap - m_min(vwap, 16)) < (m_corr(vwap, m_avg(volume, 180), 18)), 1, 0)
    AS alpha_a191_f0154
FROM data_base
"""
dai.query(get_a191_sql_base(sql_alpha_154), filters={'date': [sd, ed]}).df().dropna()

date instrument  alpha_a191_f0154
5067    2023-01-04  000001.SZ               0.0
5068    2023-01-04  000002.SZ               0.0
5069    2023-01-04  000004.SZ               0.0
5070    2023-01-04  000005.SZ               0.0
5071    2023-01-04  000006.SZ               0.0
...            ...        ...               ...
3858343 2025-12-26  688805.SH               0.0
3858344 2025-12-26  688807.SH               0.0
3858345 2025-12-26  688819.SH               0.0
3858346 2025-12-26  688981.SH               0.0
3858347 2025-12-26  689009.SH               0.0

[3668803 rows x 3 columns]

## A191 Factor 155

Formula:

`SMA(VOLUME,13,2)-SMA(VOLUME,27,2)-SMA(SMA(VOLUME,13,2)-SMA(VOLUME,27,2),10,2)`


In [19]:
sql_alpha_155 = f"""--sql
SELECT
    date,
    instrument,
    m_ta_ewm(volume, 13, 2) -
    m_ta_ewm(volume, 27, 2) -
    m_ta_ewm(
        m_ta_ewm(volume, 13, 2) - m_ta_ewm(volume, 27, 2),
        10,
        2
    )
    AS alpha_a191_f0155
FROM data_base
"""
dai.query(get_a191_sql_base(sql_alpha_155), filters={'date': [sd, ed]}).df().dropna()

date instrument  alpha_a191_f0155
5067    2023-01-04  000001.SZ      0.000000e+00
5068    2023-01-04  000002.SZ      0.000000e+00
5069    2023-01-04  000004.SZ      0.000000e+00
5070    2023-01-04  000005.SZ      0.000000e+00
5071    2023-01-04  000006.SZ      0.000000e+00
...            ...        ...               ...
3858343 2025-12-26  688805.SH     -2.546148e+05
3858344 2025-12-26  688807.SH     -1.964832e+05
3858345 2025-12-26  688819.SH      5.177559e+04
3858346 2025-12-26  688981.SH      2.327085e+06
3858347 2025-12-26  689009.SH      3.718422e+05

[3668803 rows x 3 columns]

## A191 Factor 156

Formula:

`(MAX(RANK(DECAYLINEAR(DELTA(VWAP, 5), 3)), RANK(DECAYLINEAR(((DELTA(((OPEN * 0.15) + (LOW *0.85)),2) / ((OPEN * 0.15) + (LOW * 0.85))) * -1), 3))) * -1)`


In [20]:
sql_alpha_156 = f"""--sql
SELECT
    date,
    instrument,
    (greatest(
        c_pct_rank(m_decay_linear(m_delta(vwap, 5), 3)),
        c_pct_rank(
            m_decay_linear(
                ((m_delta(((open * 0.15) + (low * 0.85)), 2) /
                  ((open * 0.15) + (low * 0.85))) * -1),
                3
            )
        )
    ) * -1)
    AS alpha_a191_f0156
FROM data_base
"""
dai.query(get_a191_sql_base(sql_alpha_156), filters={'date': [sd, ed]}).df().dropna()

date instrument  alpha_a191_f0156
25337   2023-01-10  000001.SZ         -0.277835
25338   2023-01-10  000002.SZ         -0.698672
25339   2023-01-10  000004.SZ         -0.841879
25340   2023-01-10  000005.SZ         -0.769356
25341   2023-01-10  000006.SZ         -0.983044
...            ...        ...               ...
3858342 2025-12-26  688802.SH         -0.996899
3858344 2025-12-26  688807.SH         -0.372093
3858345 2025-12-26  688819.SH         -0.691487
3858346 2025-12-26  688981.SH         -0.984099
3858347 2025-12-26  689009.SH         -0.923837

[3647592 rows x 3 columns]

## A191 Factor 157

Formula:

`(MIN(PROD(RANK(RANK(LOG(SUM(TSMIN(RANK(RANK((-1 * RANK(DELTA((CLOSE -1), 5))))), 2), 1)))), 1), 5) +TSRANK(DELAY((-1 * RET), 6), 5))`


In [21]:
sql_alpha_157 = f"""--sql
SELECT
    date,
    instrument,
    (m_min(
        m_product(
            c_pct_rank(
                c_pct_rank(
                    log(
                        m_sum(
                            m_min(
                                c_pct_rank(
                                    c_pct_rank((-1 * c_pct_rank(m_delta((close - 1), 5))))
                                ),
                                2
                            ),
                            1
                        )
                    )
                )
            ),
            1
        ),
        5
    ) + m_rank(m_lag((-1 * ret), 6), 5))
    AS alpha_a191_f0157
FROM data_base
"""
dai.query(get_a191_sql_base(sql_alpha_157), filters={'date': [sd, ed]}).df().dropna()

date instrument  alpha_a191_f0157
55750   2023-01-18  000001.SZ          5.005926
55751   2023-01-18  000002.SZ          3.992437
55752   2023-01-18  000004.SZ          3.272021
55753   2023-01-18  000005.SZ          5.405356
55754   2023-01-18  000006.SZ          3.986718
...            ...        ...               ...
3858340 2025-12-26  688799.SH          3.211874
3858341 2025-12-26  688800.SH          3.038193
3858345 2025-12-26  688819.SH          5.411012
3858346 2025-12-26  688981.SH          3.070971
3858347 2025-12-26  689009.SH          4.253155

[3615806 rows x 3 columns]

## A191 Factor 158

Formula:

`((HIGH-SMA(CLOSE,15,2))-(LOW-SMA(CLOSE,15,2)))/CLOSE`


In [22]:
sql_alpha_158 = f"""--sql
SELECT
    date,
    instrument,
    ((high - m_ta_ewm(close, 15, 2)) -
     (low - m_ta_ewm(close, 15, 2))) / close
    AS alpha_a191_f0158
FROM data_base
"""
dai.query(get_a191_sql_base(sql_alpha_158), filters={'date': [sd, ed]}).df().dropna()

date instrument  alpha_a191_f0158
5067    2023-01-04  000001.SZ          0.055168
5068    2023-01-04  000002.SZ          0.063450
5069    2023-01-04  000004.SZ          0.021084
5070    2023-01-04  000005.SZ          0.021390
5071    2023-01-04  000006.SZ          0.066558
...            ...        ...               ...
3858343 2025-12-26  688805.SH          0.073550
3858344 2025-12-26  688807.SH          0.105662
3858345 2025-12-26  688819.SH          0.027345
3858346 2025-12-26  688981.SH          0.029730
3858347 2025-12-26  689009.SH          0.039740

[3668803 rows x 3 columns]

## A191 Factor 159

Formula:

`((CLOSE-SUM(MIN(LOW,DELAY(CLOSE,1)),6))/SUM(MAX(HGIH,DELAY(CLOSE,1))-MIN(LOW,DELAY(CLOSE,1)),6)*12*24+(CLOSE-SUM(MIN(LOW,DELAY(CLOSE,1)),12))/SUM(MAX(HGIH,DELAY(CLOSE,1))-MIN(LOW,DELAY(CLOSE,1)),12)*6*24+(CLOSE-SUM(MIN(LOW,DELAY(CLOSE,1)),24))/SUM(MAX(HGIH,DELAY(CLOSE,1))-MIN(LOW,DELAY(CLOSE,1)),24)*6*24)*100/(6*12+6*24+12*24)`


In [23]:
sql_alpha_159 = f"""--sql
SELECT
    date,
    instrument,
    (
        (close - m_sum(least(low, m_lag(close, 1)), 6)) /
        m_sum(greatest(high, m_lag(close, 1)) - least(low, m_lag(close, 1)), 6) * 12 * 24 +
        (close - m_sum(least(low, m_lag(close, 1)), 12)) /
        m_sum(greatest(high, m_lag(close, 1)) - least(low, m_lag(close, 1)), 12) * 6 * 24 +
        (close - m_sum(least(low, m_lag(close, 1)), 24)) /
        m_sum(greatest(high, m_lag(close, 1)) - least(low, m_lag(close, 1)), 24) * 6 * 24
    ) * 100 / (6 * 12 + 6 * 24 + 12 * 24)
    AS alpha_a191_f0159
FROM data_base
"""
dai.query(get_a191_sql_base(sql_alpha_159), filters={'date': [sd, ed]}).df().dropna()

date instrument  alpha_a191_f0159
121771  2023-02-13  000001.SZ      -3730.867635
121772  2023-02-13  000002.SZ      -3439.193140
121773  2023-02-13  000004.SZ      -3843.982373
121774  2023-02-13  000005.SZ      -5898.670475
121775  2023-02-13  000006.SZ      -2873.151765
...            ...        ...               ...
3858340 2025-12-26  688799.SH      -3117.316602
3858341 2025-12-26  688800.SH      -1590.770662
3858345 2025-12-26  688819.SH      -5010.236847
3858346 2025-12-26  688981.SH      -3031.514916
3858347 2025-12-26  689009.SH      -4381.796685

[3547007 rows x 3 columns]

## A191 Factor 160

Formula:

`SMA((CLOSE<=DELAY(CLOSE,1)?STD(CLOSE,20):0),20,1)`


In [24]:
sql_alpha_160 = f"""--sql
SELECT
    date,
    instrument,
    m_ta_ewm(IF(close <= m_lag(close, 1), m_stddev(close, 20), 0), 20, 1)
    AS alpha_a191_f0160
FROM data_base
"""
dai.query(get_a191_sql_base(sql_alpha_160), filters={'date': [sd, ed]}).df().dropna()

date instrument  alpha_a191_f0160
5067    2023-01-04  000001.SZ          0.000000
5068    2023-01-04  000002.SZ          0.000000
5069    2023-01-04  000004.SZ          0.000000
5070    2023-01-04  000005.SZ          0.000000
5071    2023-01-04  000006.SZ          0.000000
...            ...        ...               ...
3858340 2025-12-26  688799.SH          1.171279
3858341 2025-12-26  688800.SH          5.152305
3858345 2025-12-26  688819.SH          0.646924
3858346 2025-12-26  688981.SH          2.298144
3858347 2025-12-26  689009.SH          0.897350

[3624326 rows x 3 columns]